### ISSUE TITLE
ops: RDS upgrade readiness — Multi-AZ, custom parameter group with logical_replication, maintenance window, EOL tracking

### ISSUE THREAD

**Issue by dev-dull** [dev-dull/AL_PHA#102]
## Why

`engine_version = "16"` in \`infra/rds.tf\` is pinned to the major version, so AWS-driven minor patches auto-apply but major-version upgrades require us to drive them. AWS provides standard support for each Postgres major version for **3+ years from RDS availability** ([AWS RDS Postgres release calendar](https://docs.aws.amazon.com/AmazonRDS/latest/PostgreSQLReleaseNotes/postgresql-release-calendar.html)). For Postgres 16 that's roughly November 2028, with Extended Support available at extra cost for ~3 years after that. There's no urgency now but the work that *needs* to be in place to do an upgrade smoothly takes weeks to ship — better to land it during the quiet period than under deprecation deadline.

This issue is the **infrastructure** prerequisite. Companion issues: #103 (RDS Proxy for connection-level resilience) and #104 (Lambda DB resilience patterns).

## What's in place vs. needed

| Concern | Today | Needed for seamless upgrade |
|---|---|---|
| Engine version pinned to major | ✅ `"16"` | Keep — but document the upgrade ladder |
| Multi-AZ standby | ❌ `multi_az = false` | ✅ Required for failover-based upgrade |
| Custom parameter group | ❌ uses default | ✅ Required for `rds.logical_replication = 1` (Blue/Green requirement) |
| Explicit maintenance window | ❌ AWS-default | ✅ Required so AWS doesn't pick a bad time |
| Backup window | ❌ AWS-default | ✅ Predictable I/O dip window |
| EOL monitoring | ❌ none | ✅ Calendar reminder or CloudWatch metric |
| `auto_minor_version_upgrade` | ❌ unset (defaults true) | ✅ Make explicit; keep true |
| `apply_immediately` | ❌ unset | ✅ Document toggle behavior |

### 1. Custom DB parameter group with `rds.logical_replication = 1`

Required for [RDS Blue/Green Deployments](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/blue-green-deployments.html) to work at all. Without it, when the time comes to do the major-version upgrade we'd have to set this parameter, reboot the DB (downtime!), then start the blue/green cutover. Pre-staging the parameter group means the upgrade itself is reboot-free.

```hcl
resource "aws_db_parameter_group" "main" {
  name   = "${var.project}-postgres16-${var.environment}"
  family = "postgres16"

  # Required for RDS Blue/Green Deployments (#N).
  parameter {
    name         = "rds.logical_replication"
    value        = "1"
    apply_method = "pending-reboot"
  }

  # Defense-in-depth: any query that takes longer than 30s is
  # almost certainly stuck; killing it frees connections for
  # other Lambdas during a failover storm.
  parameter {
    name  = "statement_timeout"
    value = "30000"  # ms
  }

  # Useful during upgrade — TCP-level idle detection so failed
  # connections from a stale PG process drop faster.
  parameter {
    name  = "tcp_keepalives_idle"
    value = "60"
  }
}

resource "aws_db_instance" "main" {
  # ... existing config ...
  parameter_group_name = aws_db_parameter_group.main.name
}
```

When we bump from Postgres 16 → 17, we'll need a `postgres17`-family parameter group too. The blue/green workflow expects the destination parameter group to exist before the green env spawns.

### 2. Multi-AZ for non-dev environments

`infra/rds.tf` already has the comment `multi_az = false # Off for dev, on for prod` but it's hardcoded. Wire it:

```hcl
multi_az = var.environment != "dev"
```

Multi-AZ costs roughly 2x the instance hours (the standby is always running), but for `db.t4g.micro` that's still pennies per day. The payoff is failover during the upgrade goes from minutes-of-downtime to seconds-of-blip.

Important: enabling Multi-AZ on an existing single-AZ instance requires a maintenance event itself. Schedule it for the maintenance window below, not `apply_immediately`.

### 3. Explicit maintenance and backup windows

```hcl
maintenance_window      = "Sun:09:00-Sun:10:00"  # 02:00-03:00 PT
backup_window           = "08:00-09:00"          # 01:00-02:00 PT
auto_minor_version_upgrade = true
apply_immediately       = false
```

`Sun:09:00 UTC` = `02:00 PT Sunday` for the user's primary timezone — minimal user impact. Backup precedes maintenance so any I/O spike from auto-vacuum during maintenance has fresh backup state.

`auto_minor_version_upgrade = true` keeps us current on the minor patches automatically — security fixes land in the maintenance window without us thinking about it. Major version is unchanged by this flag.

`apply_immediately = false` means non-immediate parameter changes wait for the maintenance window. Reduces the chance of an accidental Friday-afternoon DB reboot.

### 4. EOL monitoring

Two reasonable approaches:

- **Calendar reminder** in the operator's calendar pointing at [AWS's release calendar](https://docs.aws.amazon.com/AmazonRDS/latest/PostgreSQLReleaseNotes/postgresql-release-calendar.html). Quarterly check. Boring, predictable, no automation overhead.
- **CloudWatch alarm** that watches the `EngineVersion` of the RDS instance and alarms when AWS publishes an EOL date within N months. Not directly supported by CloudWatch metrics — would need a Lambda that polls the AWS price-list API or pulls from `endoflife.date`'s [public JSON feed](https://endoflife.date/api/amazon-rds-postgresql.json) and emits a metric.

Recommendation: **calendar reminder for now**, automate when we have more than one DB instance. The Lambda approach is cute but the engineering cost outweighs the value at single-DB scale.

### 5. Pre-upgrade snapshot policy

Belt-and-suspenders for the major-version upgrade itself: take a manual snapshot before kicking off the blue/green deployment, even though AWS does an automatic one. Costs a few cents in S3 standard for a few days; rolling back from a manual snapshot is faster than from an automatic one because we control the lifecycle.

Document in the upgrade runbook (next section).

## Upgrade runbook (operational, not Terraform)

Add `docs/infra/rds-upgrade-runbook.md` with the step-by-step procedure for both minor and major upgrades. Skeleton:

1. **Pre-flight**: confirm parameter group, multi-AZ, backups. Snapshot tag = `pre-upgrade-vN-vN+1-YYYY-MM-DD`. Verify CloudWatch alarms from #65 are green.
2. **Dev first**: full upgrade dry-run in `dev` environment. Run the audit script + `flutter test` against it. Document any incompatibilities.
3. **Stage**: same in stage (once #68 lands).
4. **Production**: schedule announcement window via #61 maintenance-mode toggle. Run blue/green deployment per the next companion issue. Cutover during the announced window. Monitor sync API error rate for 24h post-cutover.
5. **Rollback plan**: if anything is wrong post-cutover, revert is via RDS snapshot restore (NOT in-place — Postgres data format changes between major versions are one-way). Maximum data loss = changes since the manual snapshot.

The runbook must be tested against `dev` at least once before the first real upgrade event. Goal: when we hit the first real EOL deadline, this is rehearsed, not improvised.

## Test plan

- [ ] Apply the parameter group changes; existing connections survive (no reboot from non-pending-reboot params).
- [ ] Reboot DB during maintenance window to pick up `rds.logical_replication = 1`; verify the param is `1` post-reboot via `SHOW rds.logical_replication;`.
- [ ] Enable Multi-AZ in dev. Confirm `Multi-AZ` shows `Yes` in console after the maintenance window.
- [ ] Force a failover via `aws rds reboot-db-instance --force-failover`; verify Lambda's connection cache reconnects without an unrecovered request.
- [ ] `statement_timeout` works: run a deliberately-long query (`SELECT pg_sleep(60)`) from a psql session, verify it's killed at 30s.
- [ ] Runbook dry-run in dev: do an actual blue/green deployment to a phantom Postgres 16 → 16 (same version) just to validate the operational procedure end-to-end.

## What this does NOT do

- **The actual major-version upgrade.** That's a planned event when EOL is in sight. This issue makes it possible to do that upgrade with the existing companion infra; the upgrade itself is its own ticket when the time comes.
- **Pg extension audit.** We don't currently use any non-default extensions (`pgvector`, `postgis`, etc.). Verify before any upgrade.
- **Schema migration testing in CI** — covered by #63.

## Related

- #61 — server-side maintenance-mode toggle to halt sync during recovery. Needed during the major-version cutover window.
- #63 — schema migration testing in CI. Adjacent: covers Drift schema; this is Postgres engine version.
- #65 — CloudWatch alarms for sync-pathology signatures. The alarms confirm a clean upgrade post-cutover.
- #66 — weekly automated snapshot-restore + sanity-check. Adjacent: this issue makes snapshots usable; #66 validates they are.
- #68 — separate dev/stage/prod environments. Needed for the dev-first upgrade testing.
- #103 (RDS Proxy companion) — connection-level resilience during the failover events this issue enables.
- #104 (Lambda DB resilience companion) — application-level resilience for those same events.
- [AWS RDS PostgreSQL release calendar](https://docs.aws.amazon.com/AmazonRDS/latest/PostgreSQLReleaseNotes/postgresql-release-calendar.html)
- [endoflife.date — Amazon RDS PostgreSQL](https://endoflife.date/amazon-rds-postgresql)
